File: LeetCode/2096. Step-By-Step Directions From a Binary Tree Node to Another/2096.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public string GetDirections(TreeNode root, int startValue, int destValue) {
        // First, find the LCA of startValue and destValue
        TreeNode lca = LCA(root, startValue, destValue);

        // Get path from LCA to startValue and destValue
        StringBuilder pathToStart = new StringBuilder();
        StringBuilder pathToDest = new StringBuilder();

        GetPath(lca, startValue, pathToStart);
        GetPath(lca, destValue, pathToDest);

        // Reverse the path to start since we need to move upwards from start to LCA
        string upwardPath = new string('U', pathToStart.Length);

        // Both paths are collected in reverse order because of the nature of recursion
        // They need to be reversed to be used correctly
        char[] startPath = pathToStart.ToString().ToCharArray();
        Array.Reverse(startPath);
        char[] destPath = pathToDest.ToString().ToCharArray();
        Array.Reverse(destPath);

        // Convert movements from start to LCA into 'U's and then add movements from LCA to dest
        return upwardPath + new string(destPath);
    }

    private TreeNode LCA(TreeNode root, int p, int q) {
        if (root == null || root.val == p || root.val == q) return root;
        TreeNode left = LCA(root.left, p, q);
        TreeNode right = LCA(root.right, p, q);
        if (left != null && right != null) return root;
        return left ?? right;
    }

    private bool GetPath(TreeNode node, int val, StringBuilder path) {
        if (node == null) return false;
        if (node.val == val) return true;

        // Try to find the path in the left or right subtree
        if (GetPath(node.left, val, path)) {
            path.Append('L'); // Append 'L' if the path is found in the left subtree
        } else if (GetPath(node.right, val, path)) {
            path.Append('R'); // Append 'R' if the path is found in the right subtree
        } else {
            return false; // If no path found, return false
        }

        return true;
    }
}

/*
public class Solution {
    public string GetDirections(TreeNode root, int startValue, int destValue) {
        List<TreeNode> sourcePath = new List<TreeNode>();
        List<TreeNode> destPath = new List<TreeNode>();
        GetPath(root, startValue, sourcePath);
        GetPath(root, destValue, destPath);
        sourcePath.Reverse();
        destPath.Reverse();

        int i=0;
        while(i<sourcePath.Count || i<destPath.Count){
            if(sourcePath[i].val != destPath[i].val){
                i--;
                break;
            }
            if(i == sourcePath.Count-1 || i == destPath.Count-1){
                break;
            }
            i++;
        }

        int sourceLen = sourcePath.Count-i-1;
        string shortestPath = "";
        StringBuilder sb = new StringBuilder("");
        int j=sourceLen;
        while(j>0){
            sb.Append("U");
            j--;
        }
        TreeNode prev = destPath[i], curr;
        for(int k=i+1;k<destPath.Count;k++){
            curr = destPath[k];
            if(prev.left == curr){
                sb.Append("L");
            }
            else{
                sb.Append("R");
            }
            prev = curr;
        }
        return sb.ToString();
    }

    public bool GetPath(TreeNode root, int target, List<TreeNode> path){
        if(root == null){
            return false;
        }
        if(root.val == target){
            path.Add(root);
            return true;
        }
        if(GetPath(root.left, target, path) || GetPath(root.right, target,path)){
            path.Add(root);
            return true;
        }
        return false;
    }
}
*/

/*
public class Solution {
    public string GetDirections(TreeNode root, int startValue, int destValue) 
    {
        var startPath = new List<char>();
        var destPath = new List<char>();

        if (!FindPath(root, startValue, startPath) || !FindPath(root, destValue, destPath))
            return string.Empty;

        int i = 0;
        while (i < startPath.Count && i < destPath.Count && startPath[i] == destPath[i])
            i++;

        var result = new StringBuilder();
        for (int j = i; j < startPath.Count; j++)
            result.Append('U');
        
        for (int j = i; j < destPath.Count; j++)
            result.Append(destPath[j]);

        return result.ToString();
    }

    private bool FindPath(TreeNode root, int value, List<char> path)
    {
        if (root == null) return false;
        if (root.val == value) return true;

        path.Add('L');
        if (FindPath(root.left, value, path)) return true;
        path.RemoveAt(path.Count - 1);

        path.Add('R');
        if (FindPath(root.right, value, path)) return true;
        path.RemoveAt(path.Count - 1);

        return false;
    }
        
        /*
        StringBuilder startPath = new();
        StringBuilder destinationPath = new();
        
        GetPath(root, startValue, startPath);
        GetPath(root, destValue, destinationPath);
        
        startPath = new StringBuilder(new string(startPath.ToString().Reverse().ToArray()));
        destinationPath = new StringBuilder(new string(destinationPath.ToString().Reverse().ToArray()));
        
        while (startPath.Length > 0 && destinationPath.Length > 0 && startPath[0] == destinationPath[0]) {
            startPath.Remove(0, 1);
            destinationPath.Remove(0, 1);
        }
        
        for (int i = 0; i < startPath.Length; ++i) {
            if (startPath[i] == 'L' || startPath[i] == 'R'){ 
                startPath[i] = 'U';
            }
        }
        
        return startPath.ToString() + destinationPath.ToString();
    }
    
    public bool GetPath(TreeNode r, int v, StringBuilder p) {
        if (r.val == v) {
            return true;
        } else if (r.left != null && GetPath(r.left, v, p)) {
            p.Append("L");
            return true;
        } else if (r.right != null && GetPath(r.right, v, p)) {
            p.Append("R");
            return true;
        } else {
            return false;
        }
    }
    */

*/

File: LeetCode/2096. Step-By-Step Directions From a Binary Tree Node to Another/2096.cpp
class Solution
{
public:
    string getDirections(TreeNode *root, int startValue, int destValue)
    {
        TreeNode *lca = findLCA(root, startValue, destValue);

        string pathToStart = "";
        string pathToDest = "";

        getPath(lca, startValue, pathToStart);
        getPath(lca, destValue, pathToDest);

        // Reverse the paths since they are built backwards (from target to LCA)
        std::reverse(pathToStart.begin(), pathToStart.end());
        std::reverse(pathToDest.begin(), pathToDest.end());

        string result = string(pathToStart.size(), 'U'); // Convert path to start into 'U's
        result += pathToDest;                            // Add path to destination
        return result;
    }

private:
    TreeNode *findLCA(TreeNode *root, int p, int q)
    {
        if (!root || root->val == p || root->val == q)
            return root;
        TreeNode *left = findLCA(root->left, p, q);
        TreeNode *right = findLCA(root->right, p, q);
        if (left && right)
            return root;
        return left ? left : right;
    }

    bool getPath(TreeNode *node, int val, string &path)
    {
        if (!node)
            return false;
        if (node->val == val)
            return true;

        if (getPath(node->left, val, path))
        {
            path.push_back('L');
            return true;
        }
        else if (getPath(node->right, val, path))
        {
            path.push_back('R');
            return true;
        }
        return false;
    }
};
/*
class Solution {
public:
    bool findPath(TreeNode* root, int tar, string &str){
        if(root == NULL) return false;
        if(root->val == tar) return true;
        bool left = findPath( root->left, tar, str);
        if(left){
            str += 'L';
            return true;
        }
        bool right = findPath( root->right, tar, str);
        if(right){
            str += 'R';
            return true;
        }
        return false;
    }
    string getDirections(TreeNode* root, int startValue, int destValue) {
       string rootToSrc="";
       findPath(root,startValue,rootToSrc);

       string rootToDst="";
       findPath(root,destValue,rootToDst);
       cout<<"src "<<rootToSrc<<endl;
       cout<<"dst "<<rootToDst<<endl;
       int i = rootToSrc.size() - 1;
        int j = rootToDst.size() - 1;
         while(i>=0 && j>=0 && rootToSrc[i] == rootToDst[j]){
            rootToSrc.pop_back();
            rootToDst.pop_back();
            i--;
            j--;
        }
        i = rootToSrc.size();
        for(int j=0; j<i; j++){
            rootToSrc[j] = 'U';
        }
        reverse(rootToDst.begin(), rootToDst.end());
        return rootToSrc + rootToDst;
    }
};
*/
/*
class Solution {
public:
    unordered_map<TreeNode *, TreeNode *> parent;
    TreeNode *start, *end;
    string getDirections(TreeNode* root, int startValue, int destValue) {
        dfs(root, startValue, destValue);

        string path1 = getPath(root, start);
        string path2 = getPath(root, end);

        int m = path1.size();
        int n = path2.size();

        int i = 0;
        while(i < min(m, n) && path1[i] == path2[i])
            ++i;

        string res;
        res.resize(m + n - 2*i, 'U');

        for(int j = m-i; i < n; ++i, ++j) {
            res[j] = path2[i];
        }

        return res;
    }

    void dfs(TreeNode* node, int startValue, int destValue) {
        if(node->val == startValue) {
            start = node;
        } else if(node->val == destValue) {
            end = node;
        }

        if(node->left) {
            parent[node->left] = node;
            dfs(node->left, startValue, destValue);
        }

        if(node->right) {
            parent[node->right] = node;
            dfs(node->right, startValue, destValue);
        }
    }

    string getPath(TreeNode* root, TreeNode *node) {
        string res;
        while(node != root) {
            TreeNode *p = parent[node];

            if(node == p->left)
                res.push_back('L');
            else
                res.push_back('R');

            node = p;
        }

        reverse(res.begin(), res.end());

        return res;
    }
};
*/